File: leetcode/0046-permutations.ts
function permute(nums: number[]): number[][] {
    const result: number[][] = [];

    const backTracking = (
        comb: number[],
        depth: number
    ): void => {
        if (depth === 0) {
            result.push([...comb]);
            return;
        }

        for (let i = 0; i < nums.length; i++) {
            if (comb.includes(nums[i])) {
                continue;
            }

            comb.push(nums[i]);
            backTracking(comb, depth - 1);
            comb.pop();
        }
    };

    backTracking([], nums.length);

    return result;
}

File: leetcode/0046-permutations.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        backTracking(result, new ArrayList<>(), nums, nums.length);
        return result;
    }

    private void backTracking(List<List<Integer>> result,
                              List<Integer> comb,
                              int[] nums,
                              int depth) {

        if (depth == 0) {
            result.add(new ArrayList<>(comb));
            return;
        }

        for (int i = 0; i < nums.length; i++) {
            if (comb.contains(nums[i])) {
                continue;
            }

            comb.add(nums[i]);
            backTracking(result, comb, nums, depth - 1);
            comb.remove(comb.size() - 1);
        }
    }
}